### DISCUSSION TITLE
Unit Tests for Scanner

### DISCUSSION THREAD

**Discussion by peckto** [Fraunhofer-AISEC/gallia#495] [Ideas]
We already have unit tests on the transport modules and some basic end-to-end tests of the scanners via the vECU.
But we are missing unit tests on the scanner level, meaning isolated tests for the functionality/logic of the scanner.
I see the following advantages by having those:
* defined behavior of the scanner
* fewer bugs
* general improved code quality

Based on the thoughts from Russ Cox (https://research.swtch.com/testing), I'm thinking about the following test concept:
* one test case is a mapping of inputs to expected output
* the input for test cases consists of scanner config and defined ECU behavior
* expected output can be:
    * the result of the scanner (e.g. identified DIDs, ...)
    * raised exception
    * triggered event (e.g. power-cycle, reconnect)

To realize this, I see the following open points

- [ ] modularization of scanner #479 
- [ ] mock interface for all dependent modules (e.g. transport, power supply, ...)
- [ ] test infrastructure
- [ ] write test cases

One essential question is, how we want to express the test case logic. 
Should this be in code, or in some kind of configuration language which gets then parsed by the test infrastructure.

A configuration language could be expressed as a table:

| Scanner | Description | config | input | output |
| ---- | ------ | ----- | ----- | ----- |
| ScanServices | Test handling of NRC | gallia-test.toml |`{'request': '27*', 'response': '7F2711'}`| `{'scan-result': {0x01: {0x27: 'serviceNotSupported'}}, 'exception': none, 'events': []}` |
| ScanServices | Test handling of Timeout | gallia-test.toml | `{'request': '27*', 'response': 'TimeoutError'}`| `{'scan-result': {0x01: {0x27: 'TimeoutError'}}, 'exception': none, 'events': ['reconnect']}` |

What do you think about this topic?

**Comment by rumpelsepp** [comment]
The infrastructure is available since https://github.com/Fraunhofer-AISEC/gallia/pull/803
